src/nvim/message.c

```diff
@@ -1060,10 +1205,29 @@ static void msg_hist_add_multihl(HlMessage msg, bool temp)
     msg_hist_temp = entry;
   }
 
-  msg_hist_len += !temp;
+  msg_hist_len += !temp && !old_message_found;
```

Comment:
lots of new flags and special conditions are being added in this PR. please find a way to reduce them.